c/include/libsbp/cpp/frame_handler.h

```diff
@@ -0,0 +1,191 @@
+/**
+ * Copyright (C) 2020 Swift Navigation Inc.
+ * Contact: Swift Navigation <dev@swiftnav.com>
+ *
+ * This source is subject to the license found in the file 'LICENSE' which must
+ * be be distributed together with this source. All other rights reserved.
+ *
+ * THIS CODE AND INFORMATION IS PROVIDED "AS IS" WITHOUT WARRANTY OF ANY KIND,
+ * EITHER EXPRESSED OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE IMPLIED
+ * WARRANTIES OF MERCHANTABILITY AND/OR FITNESS FOR A PARTICULAR PURPOSE.
+ */
+
+#ifndef SBP_CPP_FRAME_HANDLER_H_
+#define SBP_CPP_FRAME_HANDLER_H_
+
+#include <cassert>
+#include <array>
+
+#include <libsbp/cpp/state.h>
+#include <libsbp/cpp/message_traits.h>
+
+namespace sbp {
+
+/**
+ * A helper function for calling a C++ object member function from a libsbp frame callback.
+ *
+ * This function is registered with libsbp as a callback and calls the `handle_sbp_frame`
+ * member function of an object from that callback passing through all of the arguments.
+ * The instance of the object is passed in via the `context` parameter.
+ *
+ * @tparam ClassT The class type to call the function on
+ *
+ * @param sender_id The decoded sender ID, is forwarded
+ * @param msg_type The message type, is forwarded
+ * @param payload_len The length of the payload, is forwarded
+ * @param payload A pointer to the payload, is forwarded
+ * @param frame_len The length of the frame, is forwarded
+ * @param frame A pointer to the frame, is forwarded
+ * @param context Pointer to an instance of `ClassT` to use
+ */
+template<typename ClassT>
+inline void sbp_frame_cb_passthrough(uint16_t sender_id, uint16_t msg_type,
+                                     uint8_t payload_len, uint8_t payload[],
+                                     uint16_t frame_len, uint8_t frame[],
+                                     void *context) {
+  assert(nullptr != context);
+
+  auto instance = static_cast<ClassT *>(context);
+  instance->handle_sbp_frame(sender_id, msg_type, payload_len, payload, frame_len, frame);
+}
+
+/**
+ * Base type for defining classes that handle specific SBP Frames
+ *
+ * Application classes should derive from this class if they wish to handle
+ * entire frames of specific SBP messages with a member function. `FrameHandler`
+ * instantiates all of the callback nodes, and registers the member function with
+ * the given `sbp_state_t`.
+ *
+ * Classes that derive from `FrameHandler` need to implement
+ *   void handle_sbp_frame(uint16_t sender_id, uint16_t msg_type,
```

Comment:
you can enclose the following in a `@code` `@endcode` and the doxygen will look nice

Reply:
Nice! I didn't know about those tags